systemtest/src/test/java/io/strimzi/systemtest/upgrade/kraft/AbstractKRaftUpgradeST.java

```diff
@@ -0,0 +1,227 @@
+                // Change kafka version of it's empty (null is for remove the version)
+                if (upgradeKafkaVersion == null) {
+                    cmdKubeClient().applyContent(KafkaUtils.changeOrRemoveKafkaInKRaft(kafkaYaml, null));
+                } else {
+                    cmdKubeClient().applyContent(KafkaUtils.changeOrRemoveKafkaConfigurationInKRaft(kafkaYaml, upgradeKafkaVersion.getVersion(), upgradeKafkaVersion.getMetadataVersion()));
+                }
+                // Wait for readiness
+                waitForReadinessOfKafkaCluster();
+            }
+        }
+    }
+
+    @Override
+    protected void waitForKafkaClusterRollingUpdate() {
+        LOGGER.info("Waiting for Kafka Pods with controller role to be rolled");
+        controllerPods = RollingUpdateUtils.waitTillComponentHasRolledAndPodsReady(TestConstants.CO_NAMESPACE, controllerSelector, 3, controllerPods);
+        LOGGER.info("Waiting for Kafka Pods with broker role to be rolled");
+        brokerPods = RollingUpdateUtils.waitTillComponentHasRolledAndPodsReady(TestConstants.CO_NAMESPACE, brokerSelector, 3, brokerPods);
+        LOGGER.info("Waiting for EO Deployment to be rolled");
+        // Check the TO and UO also got upgraded
+        eoPods = DeploymentUtils.waitTillDepHasRolled(TestConstants.CO_NAMESPACE, KafkaResources.entityOperatorDeploymentName(clusterName), 1, eoPods);
+    }
+
+    @Override
+    protected void waitForReadinessOfKafkaCluster() {
+        LOGGER.info("Waiting for Kafka Pods with controller role to be ready");
+        RollingUpdateUtils.waitForComponentAndPodsReady(TestConstants.CO_NAMESPACE, controllerSelector, 3);
+        LOGGER.info("Waiting for Kafka Pods with broker role to be ready");
+        RollingUpdateUtils.waitForComponentAndPodsReady(TestConstants.CO_NAMESPACE, brokerSelector, 3);
+        LOGGER.info("Waiting for EO Deployment");
+        DeploymentUtils.waitForDeploymentAndPodsReady(TestConstants.CO_NAMESPACE, KafkaResources.entityOperatorDeploymentName(clusterName), 1);
+    }
+
+    protected void changeKafkaAndMetadataVersion(CommonVersionModificationData versionModificationData, ExtensionContext extensionContext) throws IOException {
+        changeKafkaAndMetadataVersion(versionModificationData, false, extensionContext);
+    }
+
+    // replaceEvenIfMissing is current workaround for the situation when `metadataVersion` is not set in Kafka CR -> that's because previous version of operator
+    // doesn't contain this kind of field, so even if we set this field in the Kafka CR, it is removed by the operator
+    // this is needed for correct functionality of the `testUpgradeAcrossVersionsWithUnsupportedKafkaVersion` test
+    @SuppressWarnings("CyclomaticComplexity")
+    protected void changeKafkaAndMetadataVersion(CommonVersionModificationData versionModificationData, boolean replaceEvenIfMissing, ExtensionContext extensionContext) throws IOException {
```

Comment:
This method is really hard to read in matter of complexity :) Wouldn't be better to somehow refactor part of this method to separate methods to make this more readable? :)

Reply:
Yeah the same one is in regular upgrade/downgrade and it's also really hard to read. I will have a look on this one and in some next PR I can change the one for regular upgrade.